### PR TITLE
Add a MEF CompositionManager.

### PR DESCRIPTION
**Pull request by KirillOsenkov**
Adds an (unused for now) CompositionManager.

### CODE REVIEW

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Composition/CompositionManager.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+			ReadAssembliesFromAddins (assemblies, "/MonoDevelop/Ide/TypeService/MefHostServices");
+
+			// spawn discovery tasks in parallel for each assembly
+			var tasks = new List<Task<DiscoveredParts>> (assemblies.Count);
+			foreach (var assembly in assemblies) {
+				var task = Task.Run (() => Discovery.CreatePartsAsync (assembly));
+				tasks.Add (task);
+			}
+
+			foreach (var task in tasks) {
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Composition/CompositionManager.cs]: Maybe it's better to use Task.WhenAny here to save some time spent

Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Composition/CompositionManager.cs]: I.e. 

```
while (list.Count > 0) {
    var task = await Task.WhenAny (list);
    catalog = catalog.AddParts (task.Result);
    list.Remove (task);
}

Comment (KirillOsenkov) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Composition/CompositionManager.cs]: I was thinking about that. If you notice in my original PR I await Task.WhenAll and then iterate.

I've thought to save some time (potentially) by not waiting on all of them to complete, and start processing the first one while the others are still running.

Yours is even better in terms of perf, but what I'm worried about is it adds non-determinism to the composition. So the composition errors we'd get (and it's inevitable, people will run into composition errors all the time as part of regular development) would be non-deterministic depending on which parts were resolved in what order.

We should measure my approach and your approach and see if the optimization you propose is worth it. The overall idea is to hide the composition in a background thread while we're busy showing the start page and doing other stuff, so not sure we even need to optimize this.

Comment (KirillOsenkov) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Composition/CompositionManager.cs]: So I've tried the original approach, my current approach and your approach, and I see no difference whatsoever on a warm filesystem.
https://github.com/KirillOsenkov/MefHost/commit/d9f3de0dc4540a9f190138f62e26e5cbe91708c2

So probably either way is OK. I'd still be curious to see if you maybe see any difference on your hardware.

Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Composition/CompositionManager.cs]: Non-determinism makes sense here, sure, not sure if the perf gain is worth it, just wondering why we did this sequentially.

There will be cases where assemblies will be inserted in deterministic (i.e. `insertbefore` in extension point), so maybe this method of handling it is better

**Pull request merged**